Add ElementSearchTree, use to speed up PointwiseInterpolator

nilsvu opened the pull request:
## Proposed changes

Uses a `boost::geometric::index::rtree` to speed up repeated interpolation to single points. This speeds up ray tracing through volume data by about a factor of 3.

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->

<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

nilsvu committed:
PointwiseInterpolator: factor out interpolation in block-logical coords

nilsvu committed:
Use ElementSearchTree to speed up PointwiseInterpolator

nilsdeppe commented on src/IO/Exporter/PointwiseInterpolator.cpp:
```diff
@@ -152,10 +152,16 @@ void interpolate_to_points_impl(
     const Mesh<Dim>& mesh, const size_t offset, const size_t length,
     const std::vector<DataVector>& tensor_data) {
-  const intrp::Irregular<Dim> interpolant(mesh, x_element_logical);
+  // Copy into DataVector can be elided by supporting interpolation of double in
+  // intrp::Irregular
+  tnsr::I<DataVector, Dim, Frame::ElementLogical> x_element_logical_dv(1_st);
+  for (size_t d = 0; d < Dim; ++d) {
+    x_element_logical_dv.get(d) = x_element_logical.get(d);
+  }
```
Just make this change. Dealing with a lot of allocations was a pain in spec and caused a lot of inefficiencies. Comments like this get lost easily.

nilsvu replied:
Done. Also added a commit to interpolate single-precision data.

nilsvu committed:
Support single target point in IrregularInterpolant

Allows to elide allocation of a DataVector sometimes.

wthrowe commented on src/Domain/Structure/ElementSearchTree.hpp:
```diff
@@ -0,0 +1,100 @@
+#include "DataStructures/Tensor/Tensor.hpp"
+#include "Domain/Structure/ElementId.hpp"
+
+namespace bg = boost::geometry;
```
This is in the global namespace in a header.  It should have a more descriptive name to avoid conflicts.  Or just remove it since it's only used 6 times.

wthrowe commented on src/Domain/ElementLogicalCoordinates.cpp:
```diff
@@ -56,7 +56,7 @@ element_logical_coordinates(
   const auto& block_search_tree = block_search_tree_it->second;
   const auto found_element_id = block_search_tree.qbegin(
-      boost::geometry::index::contains(block_logical_coords.data));
+      boost::geometry::index::covers(block_logical_coords.data));
```
Is this a bug in the previous commit?  Fix it there.

nilsdeppe commented on src/Domain/Structure/ElementSearchTree.hpp:
```diff
@@ -0,0 +1,100 @@
+
+namespace domain {
+
+template <size_t Dim>
+using ElementSearchTree =
+    bg::index::rtree<ElementId<Dim>, bg::index::quadratic<16>>;
```
Why not use a strong typedef and, regardless, document this including why quadratic and 16 elements?

nilsdeppe commented on src/Domain/Structure/ElementSearchTree.hpp:
```diff
@@ -0,0 +1,100 @@
+    bg::index::rtree<ElementId<Dim>, bg::index::quadratic<16>>;
+
+/*!
+ * \brief Sorts element IDs into a search tree for efficient searching.
```
I have no idea what the input or output is. What is the size_t in the map? Why a map? It's not at all clear to me from this documentation how this would change lookup time of some spatial coordinate (x,y,z).

nilsdeppe commented on src/Domain/Structure/ElementSearchTree.hpp:
```diff
@@ -0,0 +1,100 @@
+namespace bg = boost::geometry;
+
+namespace boost::geometry::traits {
+// Make Tensor compatible with Boost.Geometry
```
Can you add some documentation of why you need to do this for Tensor? ElementID seems "obvious"

nilsdeppe commented on src/Domain/ElementLogicalCoordinates.hpp:
```diff
@@ -39,6 +41,29 @@ element_logical_coordinates(
     const tnsr::I<double, Dim, Frame::BlockLogical>& x_block_logical,
     const ElementId<Dim>& element_id);
 
+/*!
+ * \brief Map block logical coordinates to element logical coordinates using an
+ * indexed set of elements
+ *
+ * The indexing is done by the `domain::index_element_ids` function, which
+ * places element IDs into a search tree for fast lookup. If the point is on a
+ * shared element boundary then no guarantee is made which of the abutting
+ * element IDs is returned (the returned element ID is deterministic but depends
+ * on the order in which the elements were inserted into the search tree). See
+ * the other function overload for handling multiple points and disambiguating
+ * points on shared element boundaries.
+ *
+ * \param block_logical_coords the block logical coordinates of the point
+ * \param search_tree the result of `domain::index_element_ids`
+ */
+template <size_t Dim>
+std::optional<
+    std::pair<ElementId<Dim>, tnsr::I<double, Dim, Frame::ElementLogical>>>
+element_logical_coordinates(
+    const IdPair<domain::BlockId, tnsr::I<double, Dim, Frame::BlockLogical>>&
+        block_logical_coords,
+    const std::map<size_t, domain::ElementSearchTree<Dim>>& search_tree);
+
```
separate commit, add test

nilsdeppe commented on tests/Unit/Domain/Test_ElementSearchTree.cpp:
```diff
@@ -0,0 +1,53 @@
+
+SPECTRE_TEST_CASE("Unit.Domain.ElementSearchTree", "[Domain][Unit]") {
+  const auto search_tree =
+      domain::index_element_ids<2>({{0, {{{1, 0}, {1, 0}}}},
+                                    {0, {{{1, 0}, {1, 1}}}},
+                                    {0, {{{1, 1}, {1, 0}}}},
+                                    {0, {{{2, 2}, {1, 1}}}},
+                                    {0, {{{2, 3}, {1, 1}}}},
+                                    {1, {{{1, 0}, {1, 0}}}}});
+  CHECK(
+      element_logical_coordinates<2>(
+          {domain::BlockId{0},
+           tnsr::I<double, 2, Frame::BlockLogical>{{{0.25, 0.5}}}},
+          search_tree) ==
+      std::make_pair(ElementId<2>{0, {{{2, 2}, {1, 1}}}},
+                     tnsr::I<double, 2, Frame::ElementLogical>{{{0.0, 0.0}}}));
+  CHECK(
+      element_logical_coordinates<2>(
+          {domain::BlockId{0},
+           tnsr::I<double, 2, Frame::BlockLogical>{{{-0.5, 0.5}}}},
+          search_tree) ==
+      std::make_pair(ElementId<2>{0, {{{1, 0}, {1, 1}}}},
+                     tnsr::I<double, 2, Frame::ElementLogical>{{{0.0, 0.0}}}));
+  CHECK(
+      element_logical_coordinates<2>(
+          {domain::BlockId{1},
+           tnsr::I<double, 2, Frame::BlockLogical>{{{-0.5, -0.5}}}},
+          search_tree) ==
+      std::make_pair(ElementId<2>{1, {{{1, 0}, {1, 0}}}},
+                     tnsr::I<double, 2, Frame::ElementLogical>{{{0.0, 0.0}}}));
+  CHECK_FALSE(element_logical_coordinates<2>(
+                  {domain::BlockId{1},
+                   tnsr::I<double, 2, Frame::BlockLogical>{{{0.5, 0.5}}}},
+                  search_tree)
+                  .has_value());
+  CHECK_FALSE(element_logical_coordinates<2>(
+                  {domain::BlockId{2},
+                   tnsr::I<double, 2, Frame::BlockLogical>{{{-0.5, -0.5}}}},
+                  search_tree)
+                  .has_value());
```
I have no idea how this is testing the result of `index_element_ids` or why it's trying to use `element_logical_coordinates` to do so. This looks like a test of `element_logical_coordinates` not `index_element_ids`

nilsdeppe commented on src/NumericalAlgorithms/Interpolation/IrregularInterpolant.hpp:
```diff
@@ -39,6 +39,9 @@ class Irregular {
   Irregular(
       const Mesh<Dim>& source_mesh,
       const tnsr::I<DataVector, Dim, Frame::ElementLogical>& target_points);
+  Irregular(
+      const Mesh<Dim>& source_mesh,
+      const tnsr::I<double, Dim, Frame::ElementLogical>& target_point);
   Irregular();
```
test

nilsdeppe commented on src/IO/Exporter/PointwiseInterpolator.cpp:
```diff
@@ -155,13 +155,7 @@ void interpolate_to_point_impl(
-    x_element_logical_dv.get(d) = x_element_logical.get(d);
-  }
-  const intrp::Irregular<Dim> interpolant(mesh, x_element_logical_dv);
+  const intrp::Irregular<Dim> interpolant(mesh, x_element_logical);
```
Yes, this elides it, but why can't you make two commits: 1. add support for double values to Interpolator 2. use ElementSearchTree in Interpolator

I'll review once the commits are organized better.

nilsdeppe commented on src/NumericalAlgorithms/Interpolation/IrregularInterpolant.hpp:
```diff
@@ -83,6 +83,8 @@ class Irregular {
                    const gsl::span<const double>& input) const;
   void interpolate(gsl::not_null<gsl::span<std::complex<double>>*> result,
                    const gsl::span<const std::complex<double>>& input) const;
+  void interpolate(gsl::not_null<gsl::span<float>*> result,
+                   const gsl::span<const float>& input) const;
```
Test, test, test

nilsvu committed:
Add element search tree

nilsvu committed:
Use ElementSearchTree in element_logical_coordinates

nilsvu committed:
Fix a bug in interpolate_to_points

Can't use destructive_resize when working with DataVector because
it doesn't preserve the leading elements when shrinking the size.

nilsvu committed:
Add some benchmarks

nilsvu committed:
Update benchmark timings

nilsvu commented:
Pushed updates, including some benchmarks (the search tree seems to eliminate almost all overhead from having high h-refinement). I omitted the single-precision commit for now (will push to a separate PR).

nilsdeppe commented:
I'm happy with the changes. My comments are addressed. Thank you!

nilsvu commented:
@wthrowe did you want to look at this again before we merge it?

wthrowe commented:
Yes.
